doc/design/to_run_stmt.md

```diff
@@ -0,0 +1,178 @@
+
+## Syntax Extension
+
+```SQL
+SELECT * FROM source_table
+TO RUN a_data_scientist/maxcompute_functions:1.0/my_func.data_preprocessor.a_python_func
+WITH param_a = value_a,
+     param_b = value_b
+INTO result_table
+```
+
+The SQLFlow statement above will call the python function `a_python_func` from the file `/run/my_func/data_preprocessor.py` inside the docker image `a_data_scientist/maxcompute_functions:1.0`. The attributes in the `WITH` clause will be passed to the python function as parameters.
+
+## Challenges
+
+- TO RUN can execute functions in Kubernetes and MaxCompute.
+- TO RUN can execute functions in both a single process and a distributed cluster.
+- TO RUN can execute functions built upon various data computing frameworks such as [Dask](https://github.com/dask/dask), [Mars](https://github.com/mars-project/mars), [Ray](https://github.com/ray-project/ray) and so on.
+
+## Design
+
+### Fix the challenges
+
+The semantics of `TO RUN a_data_scientist/maxcompute_functions:1.0/my_func.data_preprocessor.a_python_func` is that `a_python_func` will be executed in a docker container for Kubernetes or a PyODPS node for MaxCompute. The implementation of `a_python_func` is fully customized by users. Because the execution environments for `a_python_func` are different between Kubernetes and MaxCompute, there are some differences considering these three challenges above.
+
+Kubernetes
+
+- Execution environment  
+  `a_python_func` is executed inside the function image `a_data_scientist/maxcompute_functions:1.0`. We can customized this image at the `docker build` stage and make sure that all the dependencies `a_python_func` need are already installed in the image.
+
+- Single process execution  
+  User writes the data transform process using pandas, numpy, tsfresh and so on inside `a_python_func`.
+
+- Distributed cluster execution  
+  User can install Dask/Mars/Ray while building the docker image. Inside the implementation of `a_python_func`, we can create a Dask/Mars/Ray cluster, build a computing graph and submit the distribution execution using the API from these packages.
+
+MaxCompute
+
+- Execution environment  
+  The packages of PyODPS node are already pre-installed and user cannot customize it. `a_python_func` can only use these pre-installed python packages.
+
+- Single process execution
+  Since PyODPS node cannot guarantee it contains all the packages we need such as `tsfresh`, `a_python_func` create a mars cluster containing only one worker and tell this worker to execute another function `another_python_func` to do the data transformation via [Mars remote API](https://github.com/mars-project/mars/issues/1227). The cluster uses our function image `a_data_scientist/maxcompute_functions:1.0`, `another_python_func` and its dependencies are already installed while building this image.
```

Comment:
Does Mars execute Docker images? Or Mars submit jobs to Kubernetes to execute Docker images? If it is the latter case, how is the relationship between the Kubernetes cluster used by Mars API and the one run SQLFlow generated steps?

Reply:
Mars submit jobs to kubernetes to execute docker image.

The docker in the generated step plays the role of the client to submit the distributed data computing job. The docker in the kubernetes cluster is the one which execute the data computing logic. Even though the roles are different, we can use the same docker image. We need install the mars package the packages we use to do the data preprocessing just like `TSFresh`.